### PR TITLE
test: use barrier in `TestInflightDispatchesMetric`

### LINKED ISSUE #109: flake: TestInflightDispatchesMetric
```
    metrics_test.go:342: 
        	Error Trace:	/home/runner/work/coder/coder/coderd/notifications/metrics_test.go:342
        	Error:      	Condition never satisfied
        	Test:       	TestInflightDispatchesMetric
```

seen here: https://github.com/coder/coder/actions/runs/11363823264/job/31608588543

### PR DESCRIPTION
**Pull request by mtojek**
Fixes: https://github.com/coder/internal/issues/109

Spotted in: https://github.com/coder/coder/actions/runs/11363823264/job/31608588543

It seems that notification messages are successfully dispatched before the test runner starts checking if they are in-flight (flaky).

```
    t.go:99: 2024-10-16 10:33:05.385 [info]  coderd.manager.notifier: started  notifier_id=c502e78d-19a4-45d9-9368-f08da485d5be
    t.go:99: 2024-10-16 10:33:05.385 [debu]  coderd.manager.notifier: dequeued messages  notifier_id=c502e78d-19a4-45d9-9368-f08da485d5be  count=2
...
    t.go:99: 2024-10-16 10:33:05.671 [debu]  coderd.manager.notifier: batch completed  notifier_id=c502e78d-19a4-45d9-9368-f08da485d5be  count=2
    t.go:99: 2024-10-16 10:33:05.672 [debu]  coderd.manager.notifier: dequeued messages  notifier_id=c502e78d-19a4-45d9-9368-f08da485d5be  count=0
```

Fixes:

```
    metrics_test.go:342: 
        	Error Trace:	/home/runner/work/coder/coder/coderd/notifications/metrics_test.go:342
        	Error:      	Condition never satisfied
        	Test:       	TestInflightDispatchesMetric
```

Resolves coder/internal#109

### CODE REVIEW

File: coderd/notifications/metrics_test.go [hunk 1]
```diff
@@ -493,27 +499,30 @@ func (u *updateSignallingInterceptor) BulkMarkNotificationMessagesFailed(ctx con
 	return u.Store.BulkMarkNotificationMessagesFailed(ctx, arg)
 }
 
-type delayingHandler struct {
+type barrierHandler struct {
 	h notifications.Handler
 
-	delay time.Duration
+	wg *sync.WaitGroup
```
Comment (johnstcn) [inline coderd/notifications/metrics_test.go:505]: suggestion: Could we use a `chan struct{}` here instead?

**Review by johnstcn — APPROVED** [review]

File: coderd/notifications/metrics_test.go (hunk 1 above)
Comment (mtojek) [reply coderd/notifications/metrics_test.go:505]: Hm.. we could but can you explain why is it recommended?

**Comment by matifali** [comment]
@mtojek is it same as https://github.com/coder/internal/issues/109?

**Comment by mtojek** [comment]
Yes.

File: coderd/notifications/metrics_test.go (hunk 1 above)
Comment (johnstcn) [reply coderd/notifications/metrics_test.go:505]: Fair enough!

**Pull request merged**